[jit] In RPC Server, handle TorchScript continuations asynchronously

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34109 [jit] In RPC Server, handle TorchScript continuations asynchronously**
* #35331 [rref] Handle exceptions returned via remote() calls

This change adds glue to GraphExecutor to give the RPC server
access to the future-based Interpreter::runAsync() api.

Previously, if a server encounted a TorchScript continuation-based block
with fork/wait, it would simply block in the server thread until the handler
completed, since it uses the synchronous Interpreter::run() api.

With the ivalue::Future returned by the Interpreter, we can run the
TorchScript code asynchronously from c++ simply by connecting its
callback to the server callback.

We add test cases to cover the new logic, both rpc_async and remote.

Differential Revision: [D20194321](https://our.internmc.facebook.com/intern/diff/D20194321/)

Commit:
[Don't review yet] Allow async torchscript in rpc calls.

Thread the Interpreter runAsync() logic through the GraphExecutor,
so that we can call future-based runAsync() in the TorchScript handlers.

Differential Revision: [D20194321](https://our.internmc.facebook.com/intern/diff/D20194321/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -240,6 +241,17 @@ def heavy_rpc_torchscript(tensor):
 
+
+@torch.jit.script
+def script_sleep_and_add(x):
```
why there is a "sleep" in the name?

jjlilley replied:
Sorry, changed the name. This was a vestige of my first attempt, where I tried a sleep, until I realize that we can't time.sleep in TorchScript.

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -50,23 +50,37 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
+      auto future = PythonRpcHandler::getInstance()
+                        .jitCompilationUnit()
+                        ->get_function(scriptCall.qualifiedName())
+                        .runAsync(stack);
```
Which thread will pick up this async run? Is it some thread from JIT thread pool?

jjlilley replied:
It will start the run inline. In most cases, it will simply complete inline as-is.

If there's a torch.jit._wait (see interpreter.cpp:1140), the stack is moved and when the continuation is unblocked, it will run in an at::launch thread.

mrshenli replied:
In that case does it mean RPC agent shutdown won't be able to track those threads? If that thread launches additional RPC or use RRef after shutdown, will the errors be propagated correctly?

mrshenli replied:
This probably should be OK, as there is a pending future for it.

jjlilley replied:
Right, I think process_group_agent's shutdown is implicitly oriented towards "io threads" vs "processing threads"

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -50,23 +50,37 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
+        try {
+          Message m = ScriptResp({future->value()}).toMessage();
+          m.setId(messageId);
+          responseFuture->markCompleted(std::move(m));
```
can we use `wrap` for this as well?

jjlilley replied:
Not really:
  - wrap() instantiates a fresh responseFuture with an already-completed value.
  - We already have an existing responseFuture and need to complete it.

We could technically make wrap fill in an existing responseFuture, though it's more efficient not to do so.

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -50,23 +50,37 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
+                        ->get_function(scriptCall.qualifiedName())
+                        .runAsync(stack);
+      if (future->completed()) {
+        return wrap(std::move(ScriptResp(future->value())).toMessage());
```
What is the benefit of separating this case out? Can we consolidate this into `addCallback`? Is the purpose to avoid an additional func invocation?

jjlilley replied:
I expect that probably the majority of the time (i.e. it's just computation, no fork/wait), future will already be completed.

It's more efficient to just instantiate a completed future than to go through the motions of making  an unfilled future, setting up the callback, completing the future, and making the followup callbacks.

mrshenli replied:
ah, got it, thanks!

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -129,9 +121,48 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
         // when this request is sent to self, the caller will set forkId using
         // rrefId (OwnerRRef does not have a forkId anyway).
-        ctx.addForkOfOwner(rrefId, forkId);
+        RRefContext::getInstance().addForkOfOwner(rrefId, forkId);
```
What is the reason for this change?

jjlilley replied:
I just reverted this aspect of the change. It's a vestige of an intermediate version, where I was placing this in a callback and wanted to minimize captures.

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -129,9 +121,48 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
+      auto responseFuture = std::make_shared<FutureMessage>();
+      future->addCallback([responseFuture, future, onCompletion]() {
+        try {
+          responseFuture->markCompleted(onCompletion(future->value()));
```
ditto, can we consolidate with future completed and incomplete cases?

jjlilley replied:
Similar response - let me know if you really don't like it, but consolidating will make the common case less efficient.

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -129,9 +121,48 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
+            stack.size() == 1,
+            "Return value of a builtin operator or a "
+            "TorchScript function should be a single IValue, got a vector of "
+            "size ",
```
this is prior to this PR, but can we merge these three lines into two?

jjlilley replied:
done

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -50,23 +50,37 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
+      auto future = PythonRpcHandler::getInstance()
+                        .jitCompilationUnit()
+                        ->get_function(scriptCall.qualifiedName())
+                        .runAsync(stack);
```
Do we need to worry about the life time of the stack? Will it be possible that the stack is used after the deletion of `scriptCall`?

jjlilley replied:
I believe we're fine, please in particular see 
torch/csrc/jit/interpreter.cpp:1143, where we end up effectively move-copying the stack over into the continuation.

mrshenli commented on torch/csrc/jit/runtime/graph_executor.cpp:
```diff
@@ -483,6 +483,37 @@ void GraphExecutorImplBase::run(Stack& stack) {
+    InterpreterState state;
+  };
+  auto frame = std::make_shared<Frame>(
+    getPlanFor(stack, GraphExecutor::getDefaultNumBailOuts()));
```
Is this how we keep the stack alive?

jjlilley replied:
No, this part does not keep the Stack alive, but rather the Code. It's not 100% clear that this was necessary, actually (and actually it would probably be ok not to heap allocate it, but rather stack alloc it and std::move to a heap allocated struct, since the internals are all pimpl-to-shared-ptr, but this felt aggressive)

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1631,3 +1643,19 @@ def test_function_not_on_callee(self):
+    def test_async_script_udf(self):
+        for i in range(1, 3):
+            future = rpc.rpc_async(
+                "worker{}".format((self.rank + 1) % self.world_size),
```
I guess there is no easy way for us to test if this is indeed launched asynchronously on the server side?

jjlilley replied:
Well, I added printfs to convince myself that this was happening, and that we were hitting all the codepaths. But...

If you have ideas for what to throw into python here (particularly that will also work in TorchScript), that would be good, but otherwise, it feels tricky.

jjlilley reviewed (commented):
Thanks for looking and for the feedback! Uploading a tweaked version

Commit:
Update on "[jit] In RPC Server, handle TorchScript continuations asynchronously"


This change adds glue to GraphExecutor to give the RPC server
access to the future-based Interpreter::runAsync() api.

Previously, if a server encounted a TorchScript continuation-based block
with fork/wait, it would simply block in the server thread until the handler
completed, since it uses the synchronous Interpreter::run() api.

With the ivalue::Future returned by the Interpreter, we can run the
TorchScript code asynchronously from c++ simply by connecting its
callback to the server callback.

We add test cases to cover the new logic, both rpc_async and remote.

Differential Revision: [D20194321](https://our.internmc.facebook.com/intern/diff/D20194321/)

[ghstack-poisoned]

xush6528 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -240,6 +241,17 @@ def heavy_rpc_torchscript(tensor):
     return 0
 
+
+@torch.jit.script
```
We should put these tests in to "torch/testing/_internal/distributed/rpc/jit/rpc_test.py".

jjlilley replied:
Sure, I'll move it.

I saw both files, but I also saw other torch.jit.script code in the rpc/rpc_test.py file, so I figured that the distinction was fuzzy.

Commit:
Update on "[jit] In RPC Server, handle TorchScript continuations asynchronously"


This change adds glue to GraphExecutor to give the RPC server
access to the future-based Interpreter::runAsync() api.

Previously, if a server encounted a TorchScript continuation-based block
with fork/wait, it would simply block in the server thread until the handler
completed, since it uses the synchronous Interpreter::run() api.

With the ivalue::Future returned by the Interpreter, we can run the
TorchScript code asynchronously from c++ simply by connecting its
callback to the server callback.

We add test cases to cover the new logic, both rpc_async and remote.

Differential Revision: [D20194321](https://our.internmc.facebook.com/intern/diff/D20194321/)

[ghstack-poisoned]

Commit:
Update on "[jit] In RPC Server, handle TorchScript continuations asynchronously"


This change adds glue to GraphExecutor to give the RPC server
access to the future-based Interpreter::runAsync() api.

Previously, if a server encounted a TorchScript continuation-based block
with fork/wait, it would simply block in the server thread until the handler
completed, since it uses the synchronous Interpreter::run() api.

With the ivalue::Future returned by the Interpreter, we can run the
TorchScript code asynchronously from c++ simply by connecting its
callback to the server callback.

We add test cases to cover the new logic, both rpc_async and remote.

Differential Revision: [D20194321](https://our.internmc.facebook.com/intern/diff/D20194321/)

[ghstack-poisoned]

mrshenli commented:
Looks great to me! Thanks @jjlilley! I am just a little timid on using a non-RPC thread.

* What's the implication on RPC debug info? Is there a way for RPC to know how many of those `at::launch` threads are used?
* Shall we add a failure test to make sure that errors are propagated correctly when the user function is executed using `at::launch` thread?

jjlilley commented:
Thanks for looking.
- I'll add an "async with exceptions" test case and upload in a bit, so that we have test coverage for that codepath.
 
- Re: Rpc Debug Info - we could add a num currently active rpc calls to process_group_agent; let me write that up. 
   I think the num threads stat is going to look less important over time. There's talk with Jerry and folks about consolidating the various thread pools anyway - ideally over time, we can get rid of intermediate non-compute blocking, consolidate the thread pools to avoid thrashing so much between different pools of threads, etc.

jjlilley commented:
@mrshenli , as an update- it was good that you pushed me to write tests for exception support.

This has been a bit hairier than expected (and exposed a few issues in the rest of the plumbing, some related to the disabled test_torchscript_function_exception test, though some of it can be deferred).

I'm getting there, though, hoping to update soon.

Commit:
Update on "[jit] In RPC Server, handle TorchScript continuations asynchronously"


This change adds glue to GraphExecutor to give the RPC server
access to the future-based Interpreter::runAsync() api.

Previously, if a server encounted a TorchScript continuation-based block
with fork/wait, it would simply block in the server thread until the handler
completed, since it uses the synchronous Interpreter::run() api.

With the ivalue::Future returned by the Interpreter, we can run the
TorchScript code asynchronously from c++ simply by connecting its
callback to the server callback.

We add test cases to cover the new logic, both rpc_async and remote.

Differential Revision: [D20194321](https://our.internmc.facebook.com/intern/diff/D20194321/)

[ghstack-poisoned]

jjlilley commented:
I think this change is ready for review again.
It will merge conflict with my 35331 change (exception propagation through rref), though de-conflicting it will be reasonable, which ever one makes more sense to land first.

Commit:
rebase on "[jit] In RPC Server, handle TorchScript continuations asynchronously"


This change adds glue to GraphExecutor to give the RPC server
access to the future-based Interpreter::runAsync() api.

Previously, if a server encounted a TorchScript continuation-based block
with fork/wait, it would simply block in the server thread until the handler
completed, since it uses the synchronous Interpreter::run() api.

With the ivalue::Future returned by the Interpreter, we can run the
TorchScript code asynchronously from c++ simply by connecting its
callback to the server callback.

We add test cases to cover the new logic, both rpc_async and remote.

Differential Revision: [D20194321](https://our.internmc.facebook.com/intern/diff/D20194321/)

[ghstack-poisoned]

Commit:
Update on "[jit] In RPC Server, handle TorchScript continuations asynchronously"


This change adds glue to GraphExecutor to give the RPC server
access to the future-based Interpreter::runAsync() api.

Previously, if a server encounted a TorchScript continuation-based block
with fork/wait, it would simply block in the server thread until the handler
completed, since it uses the synchronous Interpreter::run() api.

With the ivalue::Future returned by the Interpreter, we can run the
TorchScript code asynchronously from c++ simply by connecting its
callback to the server callback.

We add test cases to cover the new logic, both rpc_async and remote.

Differential Revision: [D20194321](https://our.internmc.facebook.com/intern/diff/D20194321/)

[ghstack-poisoned]

jjlilley commented:
Are there others who I should add to this code review? Thanks!

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -178,39 +198,62 @@ void RequestCallbackImpl::processRpc(
-      // TODO: make this asynchronous
-      // scriptRemoteCall is only alive within this block, use reference to
+      auto scriptRemoteCallDone =
+          [ownerRRef, rrefId, forkId, messageId, responseFuture](IValue res) {
```
looks like we no longer need the ownerRRef ptr any more, can we move here?

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -178,39 +198,62 @@ void RequestCallbackImpl::processRpc(
+              try {
+                scriptRemoteCallDone(jitFuture->value());
+              } catch (const std::exception& e) {
+                // TODO: the better approach here would be to allow ownerRRef to
```
Since this is added in #35331, do we need to update this accordingly? I am OK with either doing it here or in a followup PR

jjlilley replied:
I'm removing this comment, thanks.

mrshenli reviewed (approved):
LGTM! Thanks!

mrshenli commented:
BTW, lint failures should have been fixed on master, could you please rebase?

jjlilley commented:
Thanks for the review! I'm just finishing running tests locally with a rebase (vs the exception change), and will upload shortly

Commit:
rebase on "[jit] In RPC Server, handle TorchScript continuations asynchronously"


This change adds glue to GraphExecutor to give the RPC server
access to the future-based Interpreter::runAsync() api.

Previously, if a server encounted a TorchScript continuation-based block
with fork/wait, it would simply block in the server thread until the handler
completed, since it uses the synchronous Interpreter::run() api.

With the ivalue::Future returned by the Interpreter, we can run the
TorchScript code asynchronously from c++ simply by connecting its
callback to the server callback.

We add test cases to cover the new logic, both rpc_async and remote.

Differential Revision: [D20194321](https://our.internmc.facebook.com/intern/diff/D20194321/)

[ghstack-poisoned]
